**Легенда**

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

### Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Загрузка библиотек, таблиц, обзор данных

In [1]:
#импортируем библиотеки

import pandas as pd
from sqlalchemy import create_engine
#from pandas import DataFrame

In [2]:
#зададим функции, который будем использовать в работе

def read_tables (table_name):
    display(table_name.head())
    table_name.describe()
    table_name.info()

In [3]:
# устанавливаем параметры

db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [4]:
# выгрузим таблицы

query = '''
            SELECT * FROM books
        '''

books = pd.io.sql.read_sql(query, con = engine)

query = '''
            SELECT * FROM authors
        '''

authors = pd.io.sql.read_sql(query, con = engine)

query = '''
            SELECT * FROM publishers
        '''

publishers = pd.io.sql.read_sql(query, con = engine)

query = '''
            SELECT * FROM ratings
        '''

ratings = pd.io.sql.read_sql(query, con = engine)

query = '''
            SELECT * FROM reviews
        '''

reviews = pd.io.sql.read_sql(query, con = engine)

In [5]:
read_tables(books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


In [6]:
read_tables(authors)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [7]:
read_tables(publishers)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


In [8]:
read_tables(ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


In [9]:
read_tables(reviews)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


**Вывод:** данные в таблицах в хорошем состоянии, пропусков нет

## Выполнение заданий

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [10]:
count_books = '''SELECT COUNT(book_id) 
                 FROM books 
                 WHERE publication_date >= '2000-01-01'
'''

pd.io.sql.read_sql(count_books, con = engine)

count
0    821

**Ответ:** после 1 января 2000 года выпущено **821** книг 

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [22]:
count_reviews = '''SELECT title, SUM(count_reviews) AS count_reviews, AVG(avg_rating) AS avg_rating
                   FROM books 
                   LEFT JOIN (SELECT book_id, COUNT(review_id) AS count_reviews
                        FROM reviews
                        GROUP BY book_id) AS a
                   ON books.book_id = a.book_id
                   LEFT JOIN (SELECT book_id, AVG(rating) AS avg_rating
                        FROM ratings 
                        GROUP BY book_id) AS b
                   ON books.book_id = b.book_id
                   GROUP BY books.title

'''

pd.io.sql.read_sql(count_reviews, con = engine)

title  count_reviews  \
0                            The Count of Monte Cristo            5.0   
1                              Count Zero (Sprawl  #2)            2.0   
2    The Botany of Desire: A Plant's-Eye View of th...            2.0   
3                                 The Poisonwood Bible            5.0   
4                                 The Canterbury Tales            3.0   
..                                                 ...            ...   
994  In the Heart of the Sea: The Tragedy of the Wh...            3.0   
995                           Of Love and Other Demons            2.0   
996               Welcome to Temptation (Dempseys  #1)            2.0   
997                      World's End (The Sandman  #8)            2.0   
998                                  Holes (Holes  #1)            5.0   

     avg_rating  
0      4.217391  
1      2.500000  
2      3.500000  
3      4.363636  
4      3.333333  
..          ...  
994    3.333333  
995    4.500000  
996    5.000000  
997    4.500000  
998    3.967742  

[999 rows x 3 columns]

**Ответ:** количество обзоров и средняя оценка книг представлены в таблице выше

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц

In [12]:
effective_publisher = ''' SELECT publisher, COUNT(book_id) as count_books 
                          FROM books LEFT JOIN publishers ON books.publisher_id = publishers.publisher_id
                          WHERE num_pages > 50
                          GROUP BY publisher
                          ORDER BY count_books DESC
                          LIMIT 1
'''

pd.io.sql.read_sql(effective_publisher, con = engine)

publisher  count_books
0  Penguin Books           42

**Ответ:** издательство **Penguin Books** выпустило больше сего книг толще 50 страниц (42 книги)

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [13]:
effective_author = '''SELECT author, ROUND(AVG(avg_rating), 1) AS avg_rating_by_author 
                      FROM
                        (SELECT author_id, title, AVG(rating) AS avg_rating, COUNT(rating) AS cnt
                         FROM books LEFT JOIN ratings ON books.book_id = ratings.book_id 
                         GROUP BY title, author_id
                         HAVING COUNT(rating_id) >= 50) AS a
                      LEFT JOIN authors ON a.author_id = authors.author_id
                      GROUP BY author
                      ORDER BY avg_rating_by_author DESC
                      LIMIT 1
'''

pd.io.sql.read_sql(effective_author, con = engine)

author  avg_rating_by_author
0  J.K. Rowling/Mary GrandPré                   4.3

**Ответ:** автор с самой высокой средней оценкой книг - **J.K. Rowling/Mary GrandPré**, оценка 4.3.

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [14]:
avg_reviews = '''SELECT ROUND(AVG(count_reviews)) AS avg_reviews
                 FROM
                    (SELECT COUNT(review_id) AS count_reviews
                     FROM
                        (SELECT username 
                         FROM ratings
                         GROUP BY username
                         HAVING COUNT(rating_id) > 50) AS a
                 LEFT JOIN reviews ON a.username = reviews.username
                 GROUP BY a.username) AS b
'''

pd.io.sql.read_sql(avg_reviews, con = engine)

avg_reviews
0         24.0

**Ответ:** среднее количество обзоров от пользователей, которые поставили больше 50 оценок, - **24**. 